pkg/inventory/inventory-client.go

```diff
@@ -24,455 +22,287 @@ import (
 
-	if dryRun.ClientOrServerDryRun() {
-		klog.V(4).Infof("dry-run create inventory object: not created")
-		return pruneIDs, nil
-	}
-	err = wrappedInv.Apply(ctx, cic.dc, cic.mapper, cic.statusPolicy)
-	return pruneIDs, err
+// UnstructuredInventory implements Inventory while also tracking the actual
+// KRM object from the cluster. This enables the client to update the
+// same object and utilize resourceVersion checks.
+type UnstructuredInventory struct {
+	BaseInventory
+	// ClusterObj represents the KRM which was last fetched from the cluster.
+	// used by the client implementation to performs updates on the object.
+	ClusterObj *unstructured.Unstructured
 }
 
-// Replace stores the passed objects in the cluster inventory object, or
-// an error if one occurred.
-func (cic *ClusterClient) Replace(ctx context.Context, localInv Info, objs object.ObjMetadataSet, status []actuation.ObjectStatus,
-	dryRun common.DryRunStrategy) error {
-	// Skip entire function for dry-run.
-	if dryRun.ClientOrServerDryRun() {
-		klog.V(4).Infoln("dry-run replace inventory object: not applied")
-		return nil
-	}
-	clusterInv, err := cic.getClusterInventoryInfo(ctx, localInv)
-	if err != nil {
-		return fmt.Errorf("failed to read inventory from cluster: %w", err)
-	}
-
-	clusterObjs, err := cic.GetClusterObjs(ctx, localInv)
-	if err != nil {
-		return fmt.Errorf("failed to read inventory objects from cluster: %w", err)
+func (ui *UnstructuredInventory) Name() string {
+	if ui.ClusterObj == nil {
+		return ""
 	}
+	return ui.ClusterObj.GetName()
+}
 
-	clusterInv, wrappedInv, err := cic.replaceInventory(clusterInv, objs, status)
-	if err != nil {
-		return err
+func (ui *UnstructuredInventory) Namespace() string {
+	if ui.ClusterObj == nil {
+		return ""
 	}
+	return ui.ClusterObj.GetNamespace()
+}
 
-	// Update not required when all objects in inventory are the same and
-	// status does not need to be updated. If status is stored, always update the
-	// inventory to store the latest status.
-	if objs.Equal(clusterObjs) && cic.statusPolicy == StatusPolicyNone {
-		return nil
+func (ui *UnstructuredInventory) ID() ID {
+	if ui.ClusterObj == nil {
+		return ""
 	}
+	// Empty string if not set.
+	return ID(ui.ClusterObj.GetLabels()[common.InventoryLabel])
+}
 
-	klog.V(4).Infof("replace cluster inventory: %s/%s", clusterInv.GetNamespace(), clusterInv.GetName())
-	klog.V(4).Infof("replace cluster inventory %d objects", len(objs))
+func (ui *UnstructuredInventory) NewEmptyInventory() Inventory {
+	return ui
+}
 
-	if err := wrappedInv.ApplyWithPrune(ctx, cic.dc, cic.mapper, cic.statusPolicy, objs); err != nil {
-		return fmt.Errorf("failed to write updated inventory to cluster: %w", err)
-	}
+var _ Inventory = &UnstructuredInventory{}
 
-	return nil
+// BaseInventory is a boilerplate struct that contains the basic methods
+// to implement Inventory. Can be extended for different inventory implementations.
+type BaseInventory struct {
```

Comment:
I was confused several times when reading this name elsewhere in the code. So I've been looking for more descriptive names.

Would InMemoryInventory be accurate? Maybe that's kind of redundant.

Do we ever use it directly or is it only every wrapped by some other implementation like UnstructuredInventory?